folding-schemes/src/folding/nova/circuits.rs

```diff
@@ -389,60 +401,38 @@ where
             cmT.clone(),
         )?;
         let r = Boolean::le_bits_to_fp_var(&r_bits)?;
-
-        // enforce that the input r_nonnat as bits matches the in-circuit computed r_bits
-        let r_nonnat_bits: Vec<Boolean<C1::ScalarField>> =
-            r_nonnat.to_bits_le()?.into_iter().take(N_BITS_RO).collect();
-        r_nonnat_bits.enforce_equal(&r_bits)?;
-
-        // Notice that NIFSGadget::verify is not checking the folding of cmE & cmW, since it will
-        // be done on the other curve.
-        let nifs_check = NIFSGadget::<C1>::verify(r, U_i.clone(), u_i.clone(), U_i1.clone())?;
-        nifs_check.conditional_enforce_equal(&Boolean::TRUE, &is_not_basecase)?;
-
-        // 4.a u_{i+1}.x[0] = H(i+1, z_0, z_i+1, U_{i+1}), this is the first output of F'
+        // Also convert r_bits to a `NonNativeFieldVar`
+        let r_nonnat = nonnative_field_var_from_le_bits(cs.clone(), &r_bits)?;
+
+        // Notice that NIFSGadget::fold_committed_instance does not fold cmE & cmW.
+        // We set `U_i1.cmE` and `U_i1.cmW` to unconstrained witnesses `U_i1_cmE` and `U_i1_cmW`
+        // respectively.
+        // The correctness of them will be checked on the other curve.
+        let mut U_i1 = NIFSGadget::<C1>::fold_committed_instance(r, U_i.clone(), u_i.clone())?;
+        U_i1.cmE = U_i1_cmE;
+        U_i1.cmW = U_i1_cmW;
+
+        // P.4.a compute and check the first output of F'
+        // Base case: u_{i+1}.x[0] == H((i+1, z_0, z_{i+1}, U_{\bot})
+        // Non-base case: u_{i+1}.x[0] == H((i+1, z_0, z_{i+1}, U_{i+1})
         let (u_i1_x, _) = U_i1.clone().hash(
             &crh_params,
             i + FpVar::<CF1<C1>>::one(),
             z_0.clone(),
             z_i1.clone(),
         )?;
-
-        u_i1_x.enforce_equal(&x)?;
+        let (u_i1_x_base, _) = CommittedInstanceVar::new_constant(cs.clone(), u_dummy)?.hash(
+            &crh_params,
+            FpVar::<CF1<C1>>::one(),
+            z_0.clone(),
+            z_i1.clone(),
+        )?;
+        let x = FpVar::new_input(cs.clone(), || Ok(self.x.unwrap_or(u_i1_x_base.value()?)))?;
```

Comment:
Should the `unwrap_or(u_i1_x_base.value())` default to `zero` instead of `u_i1_x_base.value()`?
Because with the current approach, when `self.x` is `None` the next check is checking if `x == u_i1_x_base`, where `x` is loaded directly from `u_i1_x_base`.

Reply:
IMHO we can pass an arbitrary value to the variable allocation function `new_{witness, input}` without affecting the circuit's soundness? What I am doing here is just creating fallback values that are always satisfiable when the inputs are `None` (e.g., those used in `Nova::init`). In `Nova::prove_step`, all the inputs are correctly assigned, and thus the `unwrap` call never returns the default value.

Let me also elaborate on my first claim above.

We build circuits in an input-agnostic way: although the circuit should accept correct inputs and reject incorrect ones, we are unaware if the user will assign correct or incorrect inputs to the circuit.

However, it would be inconvenient if the circuit developer needs to separate the circuit implementation and input assignment, since the developer needs to figure out which input corresponds to which gate. I guess that's why circuit-building libraries allow us to specify the value of an input. But this does not mean the value we specified in `new_{witness, input}` is what the user supplies. For a valid proof, the user has to provide correct values, even if we pass incorrect values to `new_{witness, input}`. Vice versa, the user can provide incorrect values to generate an invalid proof, even if we always pass the correct ones.

I hope this makes sense, though it is also fine for me to change it to `unwrap_or_default()` 😄

Reply:
Right, I see, makes total sense. Thanks for the explanation! ^^